contents/revisionControl/bestPracticesGit.md

```diff
@@ -0,0 +1,259 @@
+
+Changes to your own branches that is used by others should obey always-functioning-commits rule to minimize surprise. This is especially important if you expect your branch to be cherry-picked by another collaborator because they require a specific bit of code that you wrote.
+
+### Stashing
+If you need to switch between branches while in the middle of developing a commit, you can use the `git stash` command instead of committing your half-done code.
+
+```bash
+# stash your work not committed to HEAD yet by pushing it onto the stash stack
+git stash
+git stash push # equivalent to git stash
+
+# restore your most recently stashed work to your current working copy
+git stash pop
+
+# acts like git stash pop, but keeps a copy of the stash in the current stash stack
+git stash apply
+
+# list all stashes in the stack
+git stash list
+```
+
+Stashes are a purely local construct and cannot be pushed to a remote repository.
+
+Read more: [git stash - Saving Changes | Atlassian Git Tutorial](https://www.atlassian.com/git/tutorials/git-stash)
+
+---
+
+## One Logical Change to One Commit
+Commits are the building blocks of a codebase; each building block should contribute exactly one useful thing, like:
+
+- adding a new function or piece of data
+- fixing one or more bugs
+    - correcting incorrect behavior
+- refactoring code or data
+    - reorganizing code
+    - removing typos
+    - formatting code
+    - changing representation of data to a different format
+
+Each _logical change in code_ should translate to exactly one commit, nothing more or less. Doing so allows you to:
+- revert a particular logical change with no/less side effects
+- easily identify bad commits that caused change in behavior
+- collaborate with others, e.g. by cherry-picking a single logical change instead of finding a bit of this commit and a bit of that commit
+- package that logical change [with a useful commit message to explain why/how you did something](#write-good-commit-messages)
+
+It may also become necessary to scope down what you may deem as a single logical change if it results in a very large commit, as that can also [introduce other problems](https://softwareengineering.stackexchange.com/a/10796). For example, implementing a single, new feature can be thought of as one logical change to the codebase, but making a pull request for a single, large commit also makes the above benefits disappear.
+
+If you are concerned about appearances, you can always opt to [hide the sausage making](#hide-the-sausage-making) to clean up your commit history.
+
+If you have made several overlapping changes on your working directory (e.g. forgot to commit, etc), you can always [perform a patch-wise stage using `git add -p`](https://stackoverflow.com/questions/1085162/commit-only-part-of-a-file-in-git).
+
+Read more:
+- [programming practices - When to commit code?](https://softwareengineering.stackexchange.com/questions/83837/when-to-commit-code)
```

Comment:
`P`rogramming practices

Reply:
I will not implement this change as it's the original title of the link, so I want to preserve it as-is.